Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -707,17 +707,33 @@ private Implementation[] GetPrioritizedImplementations(Program program)
         old.Cancel();
       }
 
-      await verifyImplementationSemaphore.WaitAsync(cts.Token);
-      ImplIdToCancellationTokenSource.AddOrUpdate(id, cts, (k, ov) => cts);
+      var result = EnqueueVerifyImplementation(processedProgram, stats, programId, er, implementation, cts.Token,
+        taskWriter);
+      var _ = result.Unwrap().ContinueWith(t =>
+      {
+        ImplIdToCancellationTokenSource.TryRemove(id, out old);
+      });
+      return result;
+    }
+
+    /// The outer task is to wait for a semaphore to let verification start
+    /// The inner task is the actual verification of the implementation
+    public async Task<Task<VerificationResult>> EnqueueVerifyImplementation(
+      ProcessedProgram processedProgram, PipelineStatistics stats,
+      string programId, ErrorReporterDelegate er, Implementation implementation,
+      CancellationToken cancellationToken,
```

Comment:
Was needed because the upper one requires a `CancellationTokenSource` and this one only has a `CancellationToken`.